### PR TITLE
[RISCV] Add scheduler model for sifive-p450.

### PR DESCRIPTION
**Pull request by topperc**
This is a slightly cleaned up version of what we've been using in our downstream toolchain.

### CODE REVIEW

File: llvm/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -153,6 +153,7 @@ Changes to the RISC-V Backend
   "SiFive Custom Instruction Extension" as SiFive defines it. The LLVM project
   needs to work with SiFive to define and document real extension names for
   individual CSRs and instructions.
+* ``-mcpu=sifive-p450`` was added.
```
Comment (wangpc-pp) [inline llvm/docs/ReleaseNotes.rst:156]: I think this patch is not for `-mcpu`, we have already added it before. This patch adds the scheduling model, and it doesn't affect any noticeable options.

File: llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+
+// SiFiveP400 machine model for scheduling and other instruction cost heuristics.
+// https://github.com/sifive/federation/blob/master/specs/design/standard/u84.json
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td:12]: This URL is 404. Is it a private repo?

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVSchedSiFiveP400.td:12]: Yes. I should remove that. It was copied from our internal repo

File: llvm/docs/ReleaseNotes.rst (hunk 1 above)
Comment (topperc) [reply llvm/docs/ReleaseNotes.rst:156]: It still needs to be in the release notes and should have been flagged in the previous review. I can commit this directly if you prefer?

Comment (wangpc-pp) [reply llvm/docs/ReleaseNotes.rst:156]: > I can commit this directly if you prefer?

Yes, that's OK.

File: llvm/lib/Target/RISCV/RISCV.td [hunk 3]
```diff
@@ -36,6 +36,7 @@ include "GISel/RISCVRegisterBanks.td"
 
 include "RISCVSchedRocket.td"
 include "RISCVSchedSiFive7.td"
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCV.td:38]: One question I should have asked before (not for this patch): Can we separate SiFive7 model into two files or two models (scalar scheduling and RVV scheduling)? I know the microarchitecture of X280 inherits from SiFive7, but there are some different designs I see from current scheduling model, which makes X280 and SiFive7 independent enough.

Maybe we can make scalar parts a multiclass and reuse it in X280 scheduling model.
cc @michaelmaitland

**Review by wangpc-pp — APPROVED** [review]
LGTM.
Though I don't know the details about microarchitecture, I believe this scheduling model has been checked internally.

File: llvm/lib/Target/RISCV/RISCV.td (hunk 3 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCV.td:38]: > but there are some different designs I see from current scheduling model, which makes X280 and SiFive7 independent enough.

I'm not sure what different designs you're referring to here?

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCV.td:38]: OK, I may use the wrong words here. I meant that the CPU resources/pipeline are different, X280 has additional vector pipeline. Intuitively, they should be separated (I think). And, the outputs of `llvm-mca` contain vector resources even for scalar CPU like `sifive-u74`.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCV.td:38]: I am currently exploring some refactors of the scheduler model in our downstream for better code reusability. I will take a look to see if what you suggest is possible. I see your point about the llvm-mca reporting. But besides llvm-mca, the suggested change doesn't impact scheduling. In my experience, the relevant TblGen classes may make this change harder than it is worth.

**Pull request merged**